### ISSUE TITLE
Allow equality comparison of Point with non-iterable objects

### ISSUE THREAD

**Issue by drdaeman** [geopy/geopy#196]
Can we please allow equality comparison with non-iterable objects, like `None`?

For example, some libraries or frameworks (Django) may consider doing something like `if value in self.empty_values` where `self.empty_values = [None]`. If `value` happens to be a `Point` instance an `TypeError` exception will be raised by `Point.__eq__`, because of `tuple(None)`. Sure, such comparison is not entirely correct, and there is a lot of ways to work around this, but this is what sometimes happens in practice, and it would be nice if things would "just work" without any extra distractions.

I'd suggest that `__eq__` should just return `False` in case the types aren't compatible (and similar logic for `__ne__` as well). I believe this would still make perfect sense, as the objects surely aren't equal.

Thanks!

**Comment by brianrusso** [comment]
100% disagree. it should not return false; throwing the exception is
correct. Simply catch the exception and handle it appropriately in your
code.

 - bri

On Mon, Jun 13, 2016 at 4:23 PM, Aleksey Zhukov <notifications@github.com>
wrote:

> Can we please allow equality comparison with non-iterable objects, like
> None?
>
> For example, some libraries or frameworks (Django) may consider doing
> something like if value in self.empty_values where self.empty_values =
> [None]. If value happens to be a Point instance an TypeError exception
> will be raised by Point.__eq__, because of tuple(None). Sure, such
> comparison is not entirely correct, and there is a lot of ways to work
> around this, but this is what sometimes happens in practice, and it would
> be nice if things would "just work" without any extra distractions.
>
> I'd suggest that __eq__ should just return False in case the types aren't
> compatible (and similar logic for __ne__ as well). I believe this would
> still make perfect sense, as the objects surely aren't equal.
>
> Thanks!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/geopy/geopy/issues/196>, or mute the thread
> <https://github.com/notifications/unsubscribe/AF_EDCtZparmp5ehMj-dIj5NF2Mh46DOks5qLbxdgaJpZM4I0szT>
> .
>



-- 
Brian Russo / (808) 271 4166

**Comment by brianrusso** [comment]
And to expand on why this is bad (returning false) - it can easily lead to
almost impossible to debug situations; where you are getting a 'false'
response, but in reality there is a type mismatch somewhere causing it to
fail 'silently'


On Mon, Jun 13, 2016 at 4:31 PM, Brian Russo <brian@beruna.org> wrote:

> 100% disagree. it should not return false; throwing the exception is
> correct. Simply catch the exception and handle it appropriately in your
> code.
>
>  - bri
>
> On Mon, Jun 13, 2016 at 4:23 PM, Aleksey Zhukov <notifications@github.com>
> wrote:
>
>> Can we please allow equality comparison with non-iterable objects, like
>> None?
>>
>> For example, some libraries or frameworks (Django) may consider doing
>> something like if value in self.empty_values where self.empty_values =
>> [None]. If value happens to be a Point instance an TypeError exception
>> will be raised by Point.__eq__, because of tuple(None). Sure, such
>> comparison is not entirely correct, and there is a lot of ways to work
>> around this, but this is what sometimes happens in practice, and it would
>> be nice if things would "just work" without any extra distractions.
>>
>> I'd suggest that __eq__ should just return False in case the types
>> aren't compatible (and similar logic for __ne__ as well). I believe this
>> would still make perfect sense, as the objects surely aren't equal.
>>
>> Thanks!
>>
>> —
>> You are receiving this because you are subscribed to this thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/geopy/geopy/issues/196>, or mute the thread
>> <https://github.com/notifications/unsubscribe/AF_EDCtZparmp5ehMj-dIj5NF2Mh46DOks5qLbxdgaJpZM4I0szT>
>> .
>>
>
>
>
> --
> Brian Russo / (808) 271 4166
>



-- 
Brian Russo / (808) 271 4166

**Comment by drdaeman** [comment]
Doesn't throwing exception here mismatches the behavior with all the standard types in Python? I mean, even completely meaningless (type-wise) comparisons work in a way I suggest `Point` to behave:

    >>> "spam" == None
    False
    >>> [] == None
    False
    >>> object() == None
    False
    >>> buffer("spam") == dict
    False
    >>> collections.namedtuple("Spam", "eggs ham")(1, 2) == (lambda x: x)
    False

Do you have any example of such situation where comparison returning `False` would lead to incorrect results?

**Comment by brianrusso** [comment]
Okay that makes a bit of sense; I misunderstood what you were asking for
earlier. A list containing None seems odd style to me personally; but I
guess there is strictly nothing wrong with it.



On Mon, Jun 13, 2016 at 5:00 PM, Aleksey Zhukov <notifications@github.com>
wrote:

> Doesn't throwing exception here mismatches the behavior with all the
> standard types in Python? I mean, even completely meaningless (type-wise)
> comparisons work in a way I suggest Point to behave:
>
> >>> "spam" == None
> False
> >>> [] == None
> False
> >>> object() == None
> False
> >>> buffer("spam") == dict
> False
> >>> collections.namedtuple("Spam", "eggs ham")(1, 2) == (lambda x: x)
> False
>
> Do you have any example of such situation where comparison returning False
> would lead to incorrect results?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/geopy/geopy/issues/196#issuecomment-225706997>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AF_EDPTMc5myPTqsYfEkXeV6rau9sEgLks5qLcT1gaJpZM4I0szT>
> .
>



-- 
Brian Russo / (808) 271 4166

**Comment by drdaeman** [comment]
Maybe it's an odd style, but this is, for example, [what Django does in its form processing modules](https://github.com/django/django/blob/267dc4ad/django/forms/fields.py#L133).

With said framework, I've tried to write a simple location form field that parses user input and returns `Point` objects, and felt uncomfortable because in my subclass I had to override a lot of methods just to handle those `TypeError`s.

The problem was, in Django's `django.forms.fields.Field` (the most basic field class) a lot of methods run checks that boils down to [`value in (None, '', [], (), {})`](https://github.com/django/django/blob/267dc4ad/django/core/validators.py#L15). I doubt if many developers would expect that `x in [None, ""]` would raise an exception, the expected behavior is probably `False` with "no, the value isn't in the list" meaning.

That's why the suggestion.
